src/ur/workday.ur

```diff
@@ -0,0 +1,378 @@
+      SendMail = bool,
+      LastUpdated = time
+]
+
+con feedbackWithoutId = [
+      Created = time,
+      RequestedDate = option time,
+      About = wid, (* who the feedback is about *)
+      Provider = wid, (* who writes the feedback *)
+      RequestedBy = option wid,
+      Body = string, (* response *)
+      Guidance = string,
+      Status = string, (* TODO: use feedbackStatus *)
+      SendMail = bool,
+      LastUpdated = time
+]
+
+(* ************************************************************************** *)
+
+functor Make(M : AUTH) = struct
+    open M
+
+    (*
+     desired features:
+      * allow manager to request feedback of their subordinates on each other
+      * notify the subordinate of this request (by email)
+      * allow the subordinate to submit the feedback
+      * notify the manager when feedback is submitted
+      * allow manager to see the feedbacks they've requested
+      * allow the subordinate to see the feedback they've been assigned
+     *)
+
+
+    val token =
+        toko <- token;
+        case toko of
+            None => error <xml>You must be logged into Workday to use this feature.</xml>
+          | Some tok => return tok
+
+    val prefixCommon = "https://wcpdev-services1.wd101.myworkday.com/ccx/api/v1/janestreet_wcpdev1"
+    val prefixPerformanceEnablement = "https://wcpdev-services1.wd101.myworkday.com/ccx/api/performanceEnablement/v5/janestreet_wcpdev1"
+    datatype service = Common | PerformanceEnablement
+    fun urlOfService service =
+        case service of
+          Common => prefixCommon
+        | PerformanceEnablement => prefixPerformanceEnablement
+
+    fun logged [a] (_ : show a) (t : transaction a) =
+        v <- t;
+        debug ("Workday response: " ^ show v);
+        return v
+
+    fun api service url =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday GET: " ^ prefix ^ url);
+          logged (WorldFfi.get (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) False)
+        end
+
+    fun apiOpt service url =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday GET: " ^ prefix ^ url);
+          logged (WorldFfi.getOpt (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) False)
+        end
+
+    fun apiPost service url body =
+        let val prefix = urlOfService service in
+          tok <- token;
+          debug ("Workday POST: " ^ prefix ^ url);
+          debug ("Workday body: " ^ body);
+          logged (WorldFfi.post (bless (prefix ^ url)) (WorldFfi.addHeader WorldFfi.emptyHeaders "Authorization" ("Bearer " ^ tok)) (Some "application/json") body)
+        end
+
+    (* NOTE: could we improve performance by using tables instead of lists? do we need to? *)
+    (* NOTE: upstream this to a world library function *)
```

Comment:
Maybe, but there are subtle pagination differences across APIs, and it's not obvious to me it is worth curating a bestiary of all the different main styles, parameterized by field names and whatever else changes between them.  This function is not long or complicated!

Reply:
Good point! Removing this comment.